### PR TITLE
Improve digitizing tools for circular strings

### PR DESCRIPTION
**Pull request by SebDieBln**
While working on #2540 I dove into the map capture tools and especially the circular string tools gave me a hard time. So I decided to go into battle and here is the result.

This PR fixes:
- The rubberband is not shown when changing from _add feature_ map tool to _add circular string by point_ map tool.
- The _add circular string by point_ map tool does not use a temporary rubberband (red dotted line).
- The _add circular string by radius_ map tool does not use a temporary rubberband (red dotted line).
- The centerpoint-rubberband (Hotkey R) does not work for the _add circular string by radius_ map tool.
- The circular string map tools do not react to the ESC key.

Furthermore some optimizations and a general cleanup of the code to make it more maintainable and reusable. But this process is not finished yet.

### CODE REVIEW

**Comment by SebDieBln** [comment]
Screenshot of a circular string maptool with one completed circular string and one being in progress, indicated by the temporary rubberband.
temprubberbandcirculartool

File: src/app/qgsmaptooladdcircularstring.cpp [hunk 1]
```diff
@@ -83,6 +86,16 @@ void QgsMapToolAddCircularString::keyPressEvent( QKeyEvent* e )
 
     createCenterPointRubberBand();
   }
+
+  if ( e && e->key() == Qt::Key_Escape )
+  {
+    mPoints.clear();
+    delete mRubberBand; mRubberBand = nullptr;
```
Comment (m-kuhn) [inline src/app/qgsmaptooladdcircularstring.cpp]: I see that there are other occurrences of this style. I think having each command on it's own line makes code much more readable and today's screens fortunately are big enough to show a couple of lines so we don't need to save vertical space ;)

**Comment by 3nids** [comment]
@m-kuhn thanks fot the assignment, but since the circular map tools are very recent, I would ask his author @mhugent to do the review...
@mhugent ok to review?

File: src/app/qgsmaptooladdcircularstring.cpp (hunk 1 above)
Comment (SebDieBln) [inline src/app/qgsmaptooladdcircularstring.cpp]: I cut-copy-pasted, but it surely is a valid point. I will add this and the others to the _code cleanup and improved readability in circular maptools_ commit on monday.

Would this be a point for the _qgisstyle_ tool? It inserts missing newlines in other situations:
```c++
if ( expression )
{
  doSomething(); }
```

Comment (m-kuhn) [inline src/app/qgsmaptooladdcircularstring.cpp]: If it can be formalized well enough into qgisstyle that would be great!

Comment (SebDieBln) [inline src/app/qgsmaptooladdcircularstring.cpp]: @m-kuhn :gift: Here you go #2598.

**Comment by mhugent** [comment]
Looks good so far. Another thing you could implement is that the circular string tools also react to the backspace key (removes the last vertex in the capture tool).

**Comment by SebDieBln** [comment]
> Another thing you could implement is that the circular string tools also react to the backspace key.

I will. But I would like to do that in a new PR that adresses some more issues, e.g. geometry validation. The key for solving these issues will be to make the code more reusable by leveraging inheritance and doing more work in the `mParentTool`, e.g. the circular tools act only as modifiers of the data in the `mParentTool`.

**Comment by SebDieBln** [comment]
@mhugent With only a week left until feature freeze, is there anything I can do to help you review and merge this PR?

**Comment by mhugent** [comment]
No problem, further enhancements can still be added in the future, so I'm merging now.

**Pull request merged**